packages/grafana-data/src/types/variables.ts

```diff
@@ -0,0 +1,93 @@
+import { ComponentType } from 'react';
+import { Observable } from 'rxjs';
+
+import {
+  DataQuery,
+  DataQueryRequest,
+  DataQueryResponse,
+  DataSourceApi,
+  DataSourceJsonData,
+  DataSourceOptionsType,
+  DataSourceQueryType,
+  QueryEditorProps,
+} from './datasource';
+
+/**
+ * Defines new variable support
+ *
+ * Use one of the properties standard | custom | datasource to define the kind of query editor and query support you want to use in the Variable system
+ *
+ * @alpha -- experimental
+ */
+export interface VariableSupport<
+  DSType extends DataSourceApi<TQuery, TOptions>,
+  TQuery extends DataQuery = DataSourceQueryType<DSType>,
+  TOptions extends DataSourceJsonData = DataSourceOptionsType<DSType>,
+  VariableQuery extends DataQuery = any
+> {
+  /**
+   * Use this property if you want to use the standard query editor supplied in the Variable system
+   */
+  standard?: StandardVariableSupport<DSType, TQuery, TOptions>;
+
+  /**
+   * Use this property if you want to use a custom query and query editor
+   */
+  custom?: CustomVariableSupport<DSType, VariableQuery, TQuery, TOptions>;
```

Comment:
@hugohaggmark would it make any sense, instead of having standard, custom and data source properties here, to have this interface defined as: 

```ts
export interface VariableSupport<
  DSType extends DataSourceApi<TQuery, TOptions>,
  TQuery extends DataQuery = DataSourceQueryType<DSType>,
  TOptions extends DataSourceJsonData = DataSourceOptionsType<DSType>,
  VariableQuery extends DataQuery = any
> {
  type: StandardVariableSupport<...> | CustomVariableSupport<...> | DataSourceVariableSupport<...>
}
```
 I imagine this will require some casting and typeguards down the road, but we would avoid complexity of deciding about the precedence of configuration. 

I'm thinking a bit from the POV of a consumer of the API who has to decide what kind of variable support he/she needs and can get confused with multiple options available as top-level properties of this interface. For me, if I don't know the underlying logic of the precedence of configured variables support, _this API is an invitation to configure more than one property_ and get surprised by the editor being chosen as the correct one in the UI.  I hope you understand what I mean.

Reply:
Yes, I totally understand what you mean and that is something I thought about. I think we would need to add a `type` prop in each sub variable support as well to make sure the guards work. I think this is worth considering, so I'll look into this next week. Thank you @dprokop!

Reply:
@dprokop I've tried out the suggested approach by you and I think it makes the api a bit clearer.